tests/pass-dep/concurrency/apple-os-unfair-lock.rs

```diff
@@ -0,0 +1,23 @@
+//@ only-target-darwin
+
+use std::cell::UnsafeCell;
+
+fn main() {
+    let lock = UnsafeCell::new(libc::OS_UNFAIR_LOCK_INIT);
+
+    unsafe {
+        libc::os_unfair_lock_lock(lock.get());
+        libc::os_unfair_lock_assert_owner(lock.get());
+        assert!(!libc::os_unfair_lock_trylock(lock.get()));
+        libc::os_unfair_lock_unlock(lock.get());
+
+        libc::os_unfair_lock_assert_not_owner(lock.get());
+    }
+
+    // `os_unfair_lock`s can be moved and leaked, even when locked,
+    // as long as no operation is occurring on the lock:
```

Comment:
No operation occurs -- where/when? Concurrently with the move? The move is just copying some bytes around, so not sure what that would even mean. Or is it about having concurrent operations on different "copies" of the same lock? Miri would currently accept that without complaining, right?

(Miri also accepts pthread locks being moved; we don't currently attempt to fully model all the UB here.)

Reply:
I meant concurrently: you shouldn't invalidate the pointer to the lock while it is being used in `os_unfair_*` in another thread.

Reply:
But if I make a copy of the lock value while it is locked, and then call `unlock` on both the old and new value -- that can't be well-defined, can it?

Reply:
Funnily enough, that actually works with the current macOS implementation, but fails in miri. I guess I could emulate that 🤷. The way `os_unfair_lock` works is that it uses the thread ID of the owning thread as a "locked"-value of an atomic variable, and then calls the macOS equivalent of futex (called `ulock`) to handle blocking and priority inheritance in the contended case. The [Fuchsia mutex](https://github.com/rust-lang/rust/blob/25acbbd12f87ed9e3eef13c095083c4693da2e8f/library/std/src/sys/sync/mutex/fuchsia.rs) in `std` (itself derived from Fuchsia's one) works exactly the same way, just with `panic!`s instead of `SIGTRAP`s.

Reply:
So... what would it do? Both unlocks succeed, and they "wake up" everyone waiting on the respective address?

No need to implement that in Miri, but having a note somewhere that we don't implement this fully correctly seems good.